new API calls for private messages in Win10 App

gerhard6380 opened the pull request:
new API calls used for Windows 10 app (similar calls to existing but
extended to include seen id and parent-uri) - would be great to include the new calls also in RC 3.5

annando commented:
Could you integrate the new data fields in "api_format_messages_win" in the standard "api_format_messages"? I don't think that we should create API calls especially for a dedicated client since these calls could be used by other clients as well. (We also shouldn't name these calls after the client)

The new fields should then start with "friendica_" like the other additional fields we already added. Additionally please update the documentation in "doc/api.md" for the new calls.

If you want to have it merged against the RC then you have to do the pull request against the RC branch. But @tobiasd and @rabuzarus should decide as well about accepting new stuff for the RC.

tobiasd commented:
>I f you want to have it merged against the RC then you have to do the pull request against the RC branch. But @tobiasd and @rabuzarus should decide as well about accepting new stuff for the RC.

I'm for it, as this PR was announced long before the freeze and delayed only because of technical difficulties with a tool used in the chain.

tobiasd commented on include/api.php:
```diff
@@ -3664,6 +3664,389 @@ function api_friendica_notification_seen($type){
+	 * @param string $type Known types are 'atom', 'rss', 'xml' and 'json'
+	 * @return string
+	 */
+	function api_friendica_direct_messages_all($type){
```
Is this function called from an API endpoint directoy? Then I would prefer if the parameters could be closer to the Twitter fuction. So the "Suffix" determines the type of the data requested.

Also what if someone has 1000000 direct messages, are they all transmitted for every call of the function? This would make usage over 2G really painful I assume.

gerhard6380 commented:
The Intention behind "api_format_messages_win" instead of changing the existing function was to avoid issues with existing clients using the standard calls as they would get a different JSON data structure back than they currently expect. However, I can change the existing function instead, if you don't see issues for existing clients. 

@tobiasd: you asked for the api_friendica_direct_messages_all(): the call is similar to the existing api/direct_messages/all - like this standard call the 'all' means inbox PLUS sent messages. And if there is no 'count' parameter given the call returns only 20 message items by default recognizing an optionally provided 'since_id' or 'max_id'.

gerhard6380 commented:
I knew that c# can jump over unknown fields in JSON, but I wasn't sure for other codes. So sounds great, I have now removed the function api_format_messages_win() and added the needed fields 'friendica_seen' and 'friendica_parent_uri' to the existing api_format_messages(). 

I will add the documentation as a next step if now the calls are ok for you.

annando commented on include/api.php:
```diff
@@ -3664,6 +3667,336 @@ function api_friendica_notification_seen($type){
+	 * @param string $type Known types are 'atom', 'rss', 'xml' and 'json'
+	 * @return string
+	 */
+	function api_friendica_direct_messages_delete($type){
```
Could you implement the Twitter standard API call for this functionality instead of creating something new?
https://dev.twitter.com/rest/reference/post/direct_messages/destroy

annando commented on include/api.php:
```diff
@@ -3664,6 +3667,336 @@ function api_friendica_notification_seen($type){
+	 * @param string $type Known types are 'atom', 'rss', 'xml' and 'json'
+	 * @return string
+	 */
+	function api_friendica_direct_messages_all($type){
```
Couldn't this call be replaced by the existing API call with maybe an additional parameter?

annando commented on include/api.php:
```diff
@@ -3664,6 +3667,336 @@ function api_friendica_notification_seen($type){
+
+
+	/**
+	 * @brief update a direct_message to seen state for Windows 10 App 
```
Please remove the "for Windows 10 App", since the API call can be used by any application.

annando commented on include/api.php:
```diff
@@ -3664,6 +3667,336 @@ function api_friendica_notification_seen($type){
+	 * @param string $type Known types are 'atom', 'rss', 'xml' and 'json'
+	 * @return string
+	 */
+	function api_friendica_direct_messages_conversation($type) {
```
Can't this function be replaced by the existing function with maybe an additional parameter of the difference to the existing function is more than the returned values?

annando commented on include/api.php:
```diff
@@ -3664,6 +3667,336 @@ function api_friendica_notification_seen($type){
+		// stop execution and return error message if no mails available
+		if($r == null) {
+			$answer = array('result' => 'error', 'message' => 'no mails available');
+			return api_format_data("direct_messages_all", $type, array('$result' => $answer));
```
The first value of "api_format_data" is the name of the root element if the output is XML. Everything that returns direct messages and is formatted like direct messages should use "direct-messages".

annando commented:
Please have a look at the Twitter documentation https://dev.twitter.com/rest/public and the GNU Social documentation https://wiki.gnusocial.de/de:gnusocial:api:twitter_compatible before adding new calls.

The goal should be to check if the intended behavior could be done with existing calls in these two APIs. If not then consider if existing calls could be extended. If the difference is only the output it should be possible to simply add the additional fields. If the difference is in the behavior then please just add an additional parameter to the call.

If you need to add a new call please always have a look at the API references to stay in the same naming scheme.

gerhard6380 commented:
Re: direct_message_delete: I have an additional parameter with the parent-uri to make sure that the call really deletes only the intended message. In opposite to Twitter I don't need the message as a returned result, but I need different status info if there are different errors on deleting. These info could be useless for Twitter clients. I don't have experience with Twitter and their clients so I would prefer letting the Twitter API call '/api/direct_message/destroy' unused for the W10 app, maybe someone else with knowledge of Twitter clients could use this in the way Twitter intends it. 

Re: direct_message_all: I have added a different return string, if there are no mails on server. This return triggers an info shown to user in Windows 10 App explaining that there are no mails at all, or he has reached the end of the list. This is not possible with the Twitter API call (both cases returns empty JSON) and adding this to the existing call could cause troubles with other clients (which I want avoid definitely).

Re: comments with 'Windows 10 App': ok, I will remove them

Re: direct_messages_conversation: you are right, this was only used to implement returning 'seen' and 'parent-uri' for the JSON output. As this is now in the standard call 'api_format_messages' I can remove my new call completely. 

Re: parameter of api_format_data: sorry I was not aware what this parameter is used for, I am not using XML output. I will change this, but just to be sure: you definitely mean 'direct-messages' and not 'direct_messages' (the dash sign!)? I have some alternative returns for my app to trigger different responses to the user. These returns contains two fields - 'result' and 'message'. Ok for you to name this with 'friendica-result' (if dash is ok, otherwise 'friendica_result') or do you have another preference?

annando commented:
Concerning "api_format_data": The first parameter is only used for XML, I think. So if your application is using the JSON ouput there is no problem. It would be best to have it named "direct-messages" like in this place: https://github.com/friendica/friendica/blob/develop/include/api.php#L2803

Concerning "direct_message_all": If the difference is only the return value then you could add a parameter like "friendica_verbose=true" to the existing call.

And with "direct_message_delete": Your system doesn't need the return value, okay. But if you would use the twitter style call then someone else could enhance the call at a later time.

I really want to avoid any redundancy of API calls. It only makes things more complicated.

gerhard6380 commented:
Re: direct_message_all: I have now included my additional code in the api_direct_message_box() and removed the separate call api_friendica_direct_messages_all. The different result when no mails are available will only be returned if the call includes 'friendica_verbose=true'.

Next I will check the delete/destroy function.

gerhard6380 commented:
Now also find the call /api/direct_messages/destroy.

annando commented:
Thank you very much for your patience! I will check the code at the evening. But question is, how to apply it to the RC branch. Hopefully @tobiasd or @rabuzarus or @fabrixxm can help with this?

Before we haven't checked this, I will not merge the pull request (because I guess that then we cannot merge it easily to the RC branch).

tobiasd commented:
@redmatrix explained some cherry oicking I'll try that after the merge

annando commented on include/api.php:
```diff
@@ -2049,6 +2049,9 @@ function api_format_messages($item, $recipient, $sender) {
 				'recipient_screen_name' => $recipient['screen_name'],
 				'sender'                => $sender,
 				'recipient'             => $recipient,
+				'title'			=> "",
```
Is "title" new? if yes, please change it to "friendica_title".

gerhard6380 replied:
No, title is not new. When you look into master branch in line 1914 you see `$ret['title'] = $item['title'] ;`. This is adding 'title' to the array as it is missing in the array definition. To keep the order (... sender, recipient, then title, then the new fields), I have added the field to the array definition.

annando replied:
Ah okay. Thanks for the clarification.

annando commented:
Besides the small annotation the pull request can be merged, I think. There are small things in the code that can be beautified, but these are no show stoppers. For example the sql queries should only select data fields that are really needed. This speeds up the calls.

rabuzarus commented on include/api.php:
```diff
@@ -2805,7 +2808,80 @@ function api_direct_messages_new($type) {
+		// optional
+		$parenturi = (x($_REQUEST, 'friendica_parenturi') ? $_REQUEST['friendica_parenturi'] : "");
+		$verbose = (x($_GET,'friendica_verbose')?strtolower($_GET['friendica_verbose']):"false");
+		// TODO: optional parameter 'include_entities' from Twitter API not yet implemented
```
please use `/// @todo` so it gets recognized by doxygen

rabuzarus commented on include/api.php:
```diff
@@ -2805,7 +2808,80 @@ function api_direct_messages_new($type) {
+				return api_format_data("direct_messages_delete", $type, array('$result' => $answer));
+			}
+		}
+		// TODO: return JSON data like Twitter API not yet implemented
```
please use /// @todo so it gets recognized by doxygen

rabuzarus commented on include/api.php:
```diff
@@ -2805,7 +2808,80 @@ function api_direct_messages_new($type) {
+			intval($id));
+	
+		// error message if specified id is not in database
+		if (count($r) == 0) {
```
`if (!dbm::is_result($r)) {`

rabuzarus commented:
@gerhard6380 
if there are db results can be checked with `if (dbm::is_result($r)) {` or `if (!dbm::is_result($r)) {`

gerhard6380 commented:
@annando: good point, I have adapted the SQL queries from 'SELECT *' to 'SELECT `id`', should improve performance as`id` is indexed. 
@rabuzarus: I have changed the todo's and the db result check to your proposals

annando commented on doc/api.md:
```diff
@@ -104,6 +104,7 @@ Unofficial Twitter command. It shows all direct answers (excluding the original
 * max_id: maximum id
 * getText: Defines the format of the status field. Can be "html" or "plain"
 * include_entities: "true" shows entities for pictures and links (Default: false)
+* friendica_verbose: "true" enables different error returns for Windows 10 app (default: "false")
```
I'm in my nitpicker mode again :-)

The "verbose" setting isn't for the windows 10 app but for every app that wants to have additional return values ;-)

annando commented:
It is now part of the "develop" branch. Hopefully with some git magic this can be applied to the RC branch as well.
